### ISSUE TITLE
Update VDM2C to use garbage collection

### ISSUE THREAD

**Issue by peterwvj** [overturetool/vdm2c#68]
The runtime library is currently being extended with support for garbage collection. The process of updating the code generator to use the garbage collector can, broadly speaking, be divided into the following tasks:

- [x] Add a `GarbageCollection` transformation that emits `*GC` calls (described below)
    - Make sure that the relevant nodes in the IR get annotated with GC related information
- [x] Re-run all existing test models with the GC option on
- [x] Add relevant garbage collection calls to the generated code (such as the "main" file)

The garbage collection extension of the runtime library currently exposes the following functions (amend the list if necessary):

- Numeric unary operators
  - [x] vdmMinusGC(TVP arg, TVP *from)
  - [x] vdmAbsGC(TVP arg, TVP *from)
  - [x] vdmFloorGC(TVP arg, TVP *from)
- Boolean operations
  - [x] vdmNotGC
  - [x] vdmOrGC
    - Added but I can't find a way to make the CG emit this.
  - [x] vdmAndGC
    - Added but I can't find a way to make the CG emit this.
  - [x] vdmImpliesGC
    - `A => B` is translated to `not A or B` so vdmImpliesGC is not used (not added).
  - [x] vdmBiimplicationGC
    - `A <=> B` is translated to `not XOR A` so vdmBiimplicationGC is not used (not added).
  - [x] vdmXorGC
    - Needed to implement vdmBiimplicationGC
- Numeric binary operators
  - [x] vdmSumGC(TVP a,TVP b, TVP *from)
  - [x] vdmDifferenceGC(TVP a,TVP b, TVP *from)
  - [x] vdmProductGC(TVP a,TVP b, TVP *from)
  - [x] vdmDivisionGC(TVP a,TVP b, TVP *from)
  - [x] vdmDivGC(TVP a, TVP b, TVP *from)
  - [x] vdmRemGC(TVP a, TVP b, TVP *from)
  - [x] vdmModGC(TVP a, TVP b, TVP *from)
  - [x] vdmPowerGC(TVP a, TVP b, TVP *from)
  - [x] vdmNumericEqualGC(TVP a, TVP b, TVP *from)
     - Not added since it is not emitted by the code generator (only used internally by the runtime)
- Numeric comparison
  - [x] vdmGreaterThanGC(TVP a, TVP b, TVP *from)
  - [x] vdmGreaterOrEqualGC(TVP a, TVP b, TVP *from)
  - [x] vdmLessThanGC(TVP a, TVP b, TVP *from)
  - [x] vdmLessOrEqualGC(TVP a, TVP b, TVP *from)
- Equality
  - [x] vdmEqualsGC(struct TypedValue* a, struct TypedValue* b, TVP *from)
- Value constructors (all handled using transformations)
  - [x] newTypeValueGC(vdmtype type, TypedValueType value, TVP *ref_from)
      - Not emitted by the code generator (only used internally by the runtime)
  - [x] newIntGC(int x, TVP *from)
  - [x] newBoolGC(bool x, TVP *from)
  - [x] newRealGC(double x, TVP *from)
  - [x] newCharGC(char x, TVP *from)
  - [x] newQuoteGC(unsigned int x, TVP *from)
- Copying
  - [x] vdmCloneGC(TVP x, TVP *from)
- Accessors
  - [x] GET_FIELD_GC(thisTypeName, fieldTypeName, classValue, fieldName)
  - [x] GET_FIELD_PTR_GC(thisTypeName, fieldTypeName, ptr, fieldName)
- Setters
  - [x] SET_FIELD_GC(thisTypeName, fieldTypeName, classValue, fieldName, newValue)
  - [x] SET_FIELD_PTR_GC(thisTypeName, fieldTypeName, ptr, fieldName, newValue)
  - [x] SET_STRUCT_FIELD_GC(tname,ptr,fieldtype,fieldname,newValue)
      - Not added since it is not emitted by the code generator (only used internally by the runtime)
- Others
  - [x] SELF_GC(objName, varName)

Note:
Strings are translated to sequences of characters so Strings do not need special attention

**Comment by peterwvj** [comment]
The test system was updated to use fixtures using the following commands

```bash
find . -name '*.cpp' | xargs sed -i -e ':a;N;$!ba;s/#include \"gtest\/gtest.h\"/#include \"gtest\/gtest.h\"\n#include \"TestFlowFunctions.h\"\n/g'
find . -name '*.cpp' | xargs sed -i 's/TEST([A-Za-z_0-9]*,/TEST_F(TestFlowFunctions,/g'
```

**Comment by peterwvj** [comment]
For practical reason we might also want to add `seqVarCG`. The problem is that currently a string is treated as sequence of characters. For example, the VDM fragment `v = "hi"` is translated to `TVP v = newSeqVar(2, newCharGC('h', NULL), newCharGC('i', NULL))`, which the garbage collector currently cannot handle. What do you think, @bandurvp ?

**Comment by peterwvj** [comment]
Currently VDM2C/GC does not consider the self expression. For example,

```
dcl tmp : SelfTest := self;
```

is currently translated to

```c
TVP tmp = newTypeValue(VDM_CLASS, (TypedValueType)
			{	.ptr=newClassValue(this->_SelfTest_id, &this->_SelfTest_refs, (freeVdmClassFunction)&SelfTest_free, this)});
```

.. but really it should be

```c
TVP tmp = newTypeValueGC(VDM_CLASS, (TypedValueType)
			{	.ptr=newClassValue(this->_SelfTest_id, &this->_SelfTest_refs, (freeVdmClassFunction)&SelfTest_free, this)}, &tmp);
```

**Comment by peterwvj** [comment]
The self expression issue was fixed in dcd52a5
